apps/language_server/lib/language_server/providers/references.ex

```diff
@@ -0,0 +1,63 @@
+defmodule ElixirLS.LanguageServer.Providers.References do
+  @moduledoc """
+  This module provides References support by using
+  the `Mix.Tasks.Xref.call/0` task to find all references to
+  any function or module identified at the provided location.
+  """
+
+  alias ElixirLS.LanguageServer.SourceFile
+  alias ElixirSense.Core.{Metadata, Parser, Source, Introspection}
+
+  def references(text, line, character, _include_declaration) do
+    xref_at_cursor(text, line, character)
+    |> Enum.filter(fn %{line: line} -> is_integer(line) end)
+    |> Enum.map(&build_location/1)
+  end
+
+  defp xref_at_cursor(text, line, character) do
+    env_at_cursor = line_environment(text, line)
+
+    subject_at_cursor(text, line, character)
+    |> Introspection.split_mod_fun_call()
+    |> expand_mod_fun(env_at_cursor)
+    |> add_arity(env_at_cursor)
+    |> callers()
+  end
+
+  defp line_environment(text, line) do
+    Parser.parse_string(text, true, true, line + 1) |> Metadata.get_env(line + 1)
+  end
+
+  defp subject_at_cursor(text, line, character) do
+    Source.subject(text, line + 1, character + 1)
+  end
+
+  defp expand_mod_fun(nil, _environment), do: nil
```

Comment:
Dialyzer says this case isn't necessary: `The pattern <'nil', _environment@1> can never match the type <{atom(),atom()},_>`